README.md

## Features

* 富文本展示组件，支持各种文字富文本，不支持图片。

* 使用LVRichText，你不需要去手动拆分字符串。一开始你就可以输入完整字符串，然后哪些子字符串需要设置成富文本就单独设置这些子字符串即可。

* LVRichTextConfig类定义了你想格式化的字符串以及配置信息

* 一定要注意，LVRichTextConfig类中的targetString既可以是单个字符串String，也可以是一个字符串数组List<String>

* 该组件内部有一些小bug，就是当一个targetString被另外一个targetString包含的时候，两者的富文本设置就会冲突，不会叠加

## Usage

```dart
          LVRichText(
          "啦啦啦蓝色加粗italic上标下标蓝色红色1红色12红色123红色红色green哈啊哈Green点击这里响应事件电话+86 18868876045网址http://www.baidu.com",
          // 设置文本的默认展示样式
          defaultTextStyle: const TextStyle(
            color: Colors.orange,
          ),
          // 不区分大小写
          caseSensitive: false,
          // 是否可以长按选中，true可以
          selectable: false,
          configList: [
            LVRichTextConfig(
              targetString: "蓝色",
              // 设置目标字符串的前后字符串，以限制匹配到指定的字符串
              stringBeforeTarget: "啦啦啦",
              stringAfterTarget: "加粗",
              style: const TextStyle(
                color: Colors.blue,
              ),
            ),
            LVRichTextConfig(
              targetString: "加粗",
              style: const TextStyle(
                fontWeight: FontWeight.bold,
              ),
            ),
            LVRichTextConfig(
              targetString: "italic",
              style: const TextStyle(
                fontStyle: FontStyle.italic,
              ),
            ),
            LVRichTextConfig(
              // 设置上下标与selectable: true冲突
              targetString: "上标",
              superScript: true,
            ),
            LVRichTextConfig(
              targetString: "下标",
              subScript: true,
            ),
            LVRichTextConfig(
              targetString: "红色",
              // 只修改匹配到的指定位置的字符串
              matchOption: LVRichTextMatchOption.specifiedIndexs,
              matchSpecifiedIndexs: [3, 4],
              style: const TextStyle(
                color: Colors.red,
              ),
            ),
            LVRichTextConfig(
              targetString: "green",
              style: const TextStyle(
                color: Colors.green,
              ),
            ),
            LVRichTextConfig(
              targetString: "点击这里响应事件",
              style: const TextStyle(
                fontWeight: FontWeight.bold,
                fontSize: 30,
              ),
              recognizer: TapGestureRecognizer()
                ..onTap = () {
                  print("点击邮箱");
                },
            ),
            // LVRichTextConfig(
            //   // 根据正则表达式去匹配字符串
            //   // 注意，直接根据正则表达式去匹配字符串的话，很容易与其他的冲突，所以尽可能避免正则与其他匹配方式一起使用
            //   targetString: "[0-9]*",
            //   style: const TextStyle(
            //     fontWeight: FontWeight.bold,
            //     fontSize: 30,
            //   ),
            // ),
            LVRichTextConfig(
                targetString: "+86 18868876045",
                // 如果targetString中包含特殊字符，则需要设置hasSpecialCharacters: true
                hasSpecialCharacters: true,
                style: const TextStyle(
                  decoration: TextDecoration.underline,
                )),
          ],
        )
```

CHANGELOG.md
## 0.0.7
* 更新示例代码

## 0.0.6
* 更新插件描述&readme文档

## 0.0.5
* 移除website/email/telphone的url校验和内部点击处理

## 0.0.4
* 移除脏数据

## 0.0.3
* CHANGELOG最新版本放在最上面
* 完善README文档
* 编写部分示例代码

## 0.0.2
* LICENSE年份改为2022

## 0.0.1
* 富文本展示组件，支持各种文字富文本，不支持图片


lib/src/rich_text.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

import 'rich_text_config.dart';

class LVRichText extends StatelessWidget {
  LVRichText(
    this.originalText, {
    Key? key,
    this.configList,
    this.defaultTextStyle,
    this.textAlign = TextAlign.start,
    this.textDirection,
    this.softWrap = true,
    this.overflow = TextOverflow.clip,
    this.maxLines,
    this.strutStyle,
    this.caseSensitive = true,
    this.selectable = false,
  });

  final String originalText;
  final List<LVRichTextConfig>? configList;
  final TextStyle? defaultTextStyle;
  final TextAlign textAlign;
  final TextDirection? textDirection;
  final bool softWrap; // 文本是否在软换行符处换行。如设为false，则代表水平方向有无限空间。
  final TextOverflow overflow;
  final int? maxLines;
  final StrutStyle? strutStyle;
  final bool caseSensitive; // 是否区分大小写，默认为true
  final bool selectable; // 是否可以长按选中

  @override
  Widget build(BuildContext context) {
    return _assembleRichText(context);
  }
}

extension LVRichTextExtensionForPrivateMethods on LVRichText {
  List<String> _specialCharacters() {
    return '\\~[]{}#%^*+=_|<>£€•.,!’()?-\$'.split('');
  }

  List<String> _processSpanTextList(
      List<LVRichTextConfig> configList, String originalText) {
    List<String> spanTextList = [];
    List<List<int>> positions = [];

    configList.asMap().forEach((index, configue) {
      String thisRegExPattern;
      String targetString = configue.targetString;
      String stringBeforeTarget = configue.stringBeforeTarget;
      String stringAfterTarget = configue.stringAfterTarget;

      bool matchLeftWordBoundary = configue.matchLeftWordBoundary;
      bool matchRightWordBoundary = configue.matchRightWordBoundary;
      bool matchWordBoundaries = configue.matchWordBoundaries;
      bool unicode = !configue.hasSpecialCharacters;

      String wordBoundaryStringBeforeTarget1 = "\\b";
      String wordBoundaryStringBeforeTarget2 = "\\s";
      String wordBoundaryStringAfterTarget1 = "\\s";
      String wordBoundaryStringAfterTarget2 = "\\b";

      String leftBoundary = "(?<!\\w)";
      String rightBoundary = "(?!\\w)";

      // matchWordBoundaries 或 matchLeftWordBoundary 中任一为 false 时，进行如下设置
      // leftBoundary = ""
      if (!matchWordBoundaries || !matchLeftWordBoundary) {
        leftBoundary = "";
        wordBoundaryStringBeforeTarget1 = "";
        wordBoundaryStringAfterTarget1 = "";
      }

      if (!matchWordBoundaries || !matchRightWordBoundary) {
        rightBoundary = "";
        wordBoundaryStringBeforeTarget2 = "";
        wordBoundaryStringAfterTarget2 = "";
      }

      bool isHan = RegExp(r"[\u4e00-\u9fa5]+",
              caseSensitive: caseSensitive, unicode: unicode)
          .hasMatch(targetString);

      bool isArabic = RegExp(r"[\u0621-\u064A]+",
              caseSensitive: caseSensitive, unicode: unicode)
          .hasMatch(targetString);

      /// 当目标字符串是汉字或者阿拉伯数字的时候，进行如下设置
      /// matchWordBoundaries = false
      /// wordBoundaryStringBeforeTarget = ""
      if (isHan || isArabic) {
        matchWordBoundaries = false;
        leftBoundary = "";
        rightBoundary = "";
        wordBoundaryStringBeforeTarget1 = "";
        wordBoundaryStringBeforeTarget2 = "";
        wordBoundaryStringAfterTarget1 = "";
        wordBoundaryStringAfterTarget2 = "";
      }

      String stringBeforeTargetRegex = "";
      if (stringBeforeTarget != "") {
        stringBeforeTargetRegex =
            "(?<=$wordBoundaryStringBeforeTarget1$stringBeforeTarget$wordBoundaryStringBeforeTarget2)";
      }
      String stringAfterTargetRegex = "";
      if (stringAfterTarget != "") {
        stringAfterTargetRegex =
            "(?=$wordBoundaryStringAfterTarget1$stringAfterTarget$wordBoundaryStringAfterTarget2)";
      }

      thisRegExPattern =
          '($stringBeforeTargetRegex$leftBoundary$targetString$rightBoundary$stringAfterTargetRegex)';
      RegExp exp = new RegExp(thisRegExPattern,
          caseSensitive: caseSensitive, unicode: unicode);
      var allMatches = exp.allMatches(originalText);

      int matchesLength = allMatches.length;
      List<int> matchIndexList = [];
      LVRichTextMatchOption matchOption = configue.matchOption;
      switch (matchOption) {
        case LVRichTextMatchOption.all:
          matchIndexList = List<int>.generate(matchesLength, (i) => i);
          break;
        case LVRichTextMatchOption.first:
          matchIndexList = [0];
          break;
        case LVRichTextMatchOption.last:
          matchIndexList = [matchesLength - 1];
          break;
        case LVRichTextMatchOption.specifiedIndexs:
          if (configue.matchSpecifiedIndexs != null) {
            configue.matchSpecifiedIndexs!.forEach((index) {
              matchIndexList.add(index);
            });
          }
          break;
      }

      // 例如：positions = [[7,11],[26,30],]
      allMatches.toList().asMap().forEach((index, match) {
        if (matchIndexList.indexOf(index) > -1) {
          positions.add([match.start, match.end]);
        }
      });
    });
    // 在某些情况下，排序结果仍然是无序的，因此需要重新排序一维列表
    positions.sort((a, b) => a[0].compareTo(b[0]));

    // 删除无效的位置
    List<List<int>> postionsToRemove = [];
    for (var i = 1; i < positions.length; i++) {
      if (positions[i][0] < positions[i - 1][1]) {
        postionsToRemove.add(positions[i]);
      }
    }
    postionsToRemove.forEach((position) {
      positions.remove(position);
    });

    //将 positions 转化成一维列表
    List<int> splitPositions = [0];
    positions.forEach((position) {
      splitPositions.add(position[0]);
      splitPositions.add(position[1]);
    });
    splitPositions.add(originalText.length);
    splitPositions.sort();

    splitPositions.asMap().forEach((index, splitPosition) {
      if (index != 0) {
        spanTextList.add(
            originalText.substring(splitPositions[index - 1], splitPosition));
      }
    });
    return spanTextList;
  }

  String _replaceSpecialCharacters(str) {
    String tempStr = str;
    //\[]()^*+?.$-{}|!
    _specialCharacters().forEach((chr) {
      tempStr = tempStr.replaceAll(chr, '\\$chr');
    });

    return tempStr;
  }

  Widget _assembleRichText(BuildContext context) {
    List<InlineSpan> textSpanList = _assembleTextSpanList(context);

    if (selectable) {
      return SelectableText.rich(
        TextSpan(
          style: defaultTextStyle ?? DefaultTextStyle.of(context).style,
          children: textSpanList,
        ),
        maxLines: maxLines,
        strutStyle: strutStyle,
        textAlign: textAlign,
        textDirection: textDirection,
      );
    } else {
      return RichText(
        text: TextSpan(
          style: defaultTextStyle ?? DefaultTextStyle.of(context).style,
          children: textSpanList,
        ),
        maxLines: maxLines,
        overflow: overflow,
        softWrap: softWrap,
        strutStyle: strutStyle,
        textAlign: textAlign,
        textDirection: textDirection,
      );
    }
  }

  List<InlineSpan> _assembleTextSpanList(BuildContext context) {
    List<InlineSpan> textSpanList = [];

    _assembleSpanTextList((spanTextList, unicode, finalConfigList) {
      spanTextList.forEach((spanText) {
        var inlineSpan;
        int targetIndex = -1;
        RegExpMatch? match;
        if (configList != null) {
          finalConfigList.asMap().forEach((index, configue) {
            String targetString = configue.targetString;

            //\$, match end
            RegExp targetStringExp = RegExp(
              '^$targetString\$',
              caseSensitive: caseSensitive,
              unicode: unicode,
            );

            RegExpMatch? tempMatch = targetStringExp.firstMatch(spanText);
            if (tempMatch is RegExpMatch) {
              targetIndex = index;
              match = tempMatch;
            }
          });
        }

        // 如果当前字符串是目标字符串
        if (targetIndex > -1) {
          var configue = finalConfigList[targetIndex];

          if (null != configue.matchBuilder && match is RegExpMatch) {
            inlineSpan = configue.matchBuilder!(context, match);
          } else if (configue.superScript && !selectable) {
            // 将目标字符串改为上标
            inlineSpan = WidgetSpan(
              child: Transform.translate(
                offset: const Offset(0, -5),
                child: Text(
                  spanText,
                  textScaleFactor: 0.7,
                  style: configue.style ?? DefaultTextStyle.of(context).style,
                ),
              ),
            );
          } else if (configue.subScript && !selectable) {
            // 将目标字符串改为下标
            inlineSpan = WidgetSpan(
              child: Transform.translate(
                offset: const Offset(0, 1),
                child: Text(
                  spanText,
                  textScaleFactor: 0.7,
                  style: configue.style ?? DefaultTextStyle.of(context).style,
                ),
              ),
            );
          } else {
            inlineSpan = TextSpan(
              text: spanText,
              recognizer: configue.recognizer,
              style: configue.style ?? DefaultTextStyle.of(context).style,
            );
          }
        } else {
          inlineSpan = TextSpan(
            text: spanText,
          );
        }
        textSpanList.add(inlineSpan);
      });
    }, context);

    return textSpanList;
  }

  _assembleSpanTextList(
    void Function(
      List<String> spanTextList,
      bool unicode,
      List<LVRichTextConfig> finalConfigList,
    )
        resultClosure,
    BuildContext context,
  ) {
    List<LVRichTextConfig> tempConfigList = [];
    List<LVRichTextConfig> finalConfigList = [];
    List<String> spanTextList = [];
    bool unicode = true;

    if (configList == null) {
      spanTextList = [originalText];
    } else {
      configList!.asMap().forEach((index, configue) {
        if (configue.targetString is List<String>) {
          // 如果数据是字符串数组
          configue.targetString.asMap().forEach((index, eachTargetString) {
            tempConfigList
                .add(configue.copyWith(targetString: eachTargetString));
          });
        } else {
          tempConfigList.add(configue);
        }
      });

      tempConfigList.asMap().forEach((index, configure) {
        if (configure.hasSpecialCharacters) {
          unicode = false;
          String newTargetString =
              _replaceSpecialCharacters(configure.targetString);
          finalConfigList
              .add(configure.copyWith(targetString: newTargetString));
        } else {
          finalConfigList.add(configure);
        }
      });

      spanTextList = _processSpanTextList(finalConfigList, originalText);
    }

    resultClosure(spanTextList, unicode, finalConfigList);
  }
}


lib/src/rich_text_config.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

enum LVRichTextMatchOption {
  all, // 匹配所有
  first, // 只匹配第一个
  last, // 只匹配最后一个
  specifiedIndexs, // 匹配指定的下标。该情况下需要配置matchSpecifiedIndexs
}

typedef LVRichTextMatchBuilder = InlineSpan Function(
    BuildContext context, RegExpMatch? match);

class LVRichTextConfig {
  final dynamic targetString; // 待格式化的数据，可以是单个字符串String，也可以是一个字符串数组List<String>
  final String stringBeforeTarget; // 待格式数据前面的字符串
  final String stringAfterTarget; // 待格式数据后面的字符串

  /*
  * 如果以下三者均设为true，则优先级为matchLeftWordBoundary > matchRightWordBoundary > matchWordBoundaries
  * 当matchWordBoundaries设置为true的时候，matchLeftWordBoundary将被设置为false
  * 当matchWordBoundaries或者matchLeftWordBoundary设置为true的时候，matchRightWordBoundary将被设置为false
  * */
  final bool matchWordBoundaries; // 是否在正则表达式中应用单词边界，默认为true
  final bool matchLeftWordBoundary; // 是否在正则中仅应用左字边界，默认为false
  final bool matchRightWordBoundary; // 是否在正则中仅应用右字边界，默认为false

  /*
  * 优先级：superScript > subScript
  * */
  final bool superScript; // 将targetString转换为上标
  final bool subScript; // 将targetString转换为下标

  final TextStyle? style;
  final GestureRecognizer? recognizer;

  final bool
      hasSpecialCharacters; // 如果targetString里面包含如下特殊字符——\~[]{}#%^*+=_|<>£€•.,!’()?-$，则需要将该属性设置为true

  final LVRichTextMatchOption matchOption; // 默认匹配全部
  final List<int>? matchSpecifiedIndexs; // matchOption为specifiedIndexs时，必须配置该值

  final LVRichTextMatchBuilder? matchBuilder;

  LVRichTextConfig({
    Key? key,
    required this.targetString,
    this.stringBeforeTarget = '',
    this.stringAfterTarget = '',
    this.matchWordBoundaries = true,
    this.matchLeftWordBoundary = true,
    this.matchRightWordBoundary = true,
    this.superScript = false,
    this.subScript = false,
    this.style,
    this.recognizer,
    this.hasSpecialCharacters = false,
    this.matchOption = LVRichTextMatchOption.all,
    this.matchSpecifiedIndexs,
    this.matchBuilder,
  });

  LVRichTextConfig copyWith({
    targetString,
    stringBeforeTarget,
    stringAfterTarget,
    matchWordBoundaries,
    matchLeftWordBoundary,
    matchRightWordBoundary,
    superScript,
    subScript,
    style,
    urlType,
    recognizer,
    hasSpecialCharacters,
    matchOption,
    matchSpecifiedIndexs,
  }) {
    return LVRichTextConfig(
      targetString: targetString ?? this.targetString,
      stringBeforeTarget: stringBeforeTarget ?? this.stringBeforeTarget,
      stringAfterTarget: stringAfterTarget ?? this.stringAfterTarget,
      matchWordBoundaries: matchWordBoundaries ?? this.matchWordBoundaries,
      matchLeftWordBoundary:
          matchLeftWordBoundary ?? this.matchLeftWordBoundary,
      matchRightWordBoundary:
          matchRightWordBoundary ?? this.matchRightWordBoundary,
      superScript: superScript ?? this.superScript,
      subScript: subScript ?? this.subScript,
      style: style ?? this.style,
      recognizer: recognizer ?? this.recognizer,
      hasSpecialCharacters: hasSpecialCharacters ?? this.hasSpecialCharacters,
      matchOption: matchOption ?? this.matchOption,
      matchSpecifiedIndexs: matchSpecifiedIndexs ?? this.matchSpecifiedIndexs,
    );
  }
}


example/lib/main.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:lavie_rich_text/src/rich_text.dart';
import 'package:lavie_rich_text/src/rich_text_config.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Rich Text Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: LVRichText(
          "啦啦啦蓝色加粗italic上标下标蓝色红色1红色12红色123红色红色green哈啊哈Green点击这里响应事件电话+86 18868876045网址http://www.baidu.com",
          // 设置文本的默认展示样式
          defaultTextStyle: const TextStyle(
            color: Colors.orange,
          ),
          // 不区分大小写
          caseSensitive: false,
          // 是否可以长按选中，true可以
          selectable: false,
          configList: [
            LVRichTextConfig(
              targetString: "蓝色",
              // 设置目标字符串的前后字符串，以限制匹配到指定的字符串
              stringBeforeTarget: "啦啦啦",
              stringAfterTarget: "加粗",
              style: const TextStyle(
                color: Colors.blue,
              ),
            ),
            LVRichTextConfig(
              targetString: "加粗",
              style: const TextStyle(
                fontWeight: FontWeight.bold,
              ),
            ),
            LVRichTextConfig(
              targetString: "italic",
              style: const TextStyle(
                fontStyle: FontStyle.italic,
              ),
            ),
            LVRichTextConfig(
              // 设置上下标与selectable: true冲突
              targetString: "上标",
              superScript: true,
            ),
            LVRichTextConfig(
              targetString: "下标",
              subScript: true,
            ),
            LVRichTextConfig(
              targetString: "红色",
              // 只修改匹配到的指定位置的字符串
              matchOption: LVRichTextMatchOption.specifiedIndexs,
              matchSpecifiedIndexs: [3, 4],
              style: const TextStyle(
                color: Colors.red,
              ),
            ),
            LVRichTextConfig(
              targetString: "green",
              style: const TextStyle(
                color: Colors.green,
              ),
            ),
            LVRichTextConfig(
              targetString: "点击这里响应事件",
              style: const TextStyle(
                fontWeight: FontWeight.bold,
                fontSize: 30,
              ),
              recognizer: TapGestureRecognizer()
                ..onTap = () {
                  print("点击邮箱");
                },
            ),
            // LVRichTextConfig(
            //   // 根据正则表达式去匹配字符串
            //   // 注意，直接根据正则表达式去匹配字符串的话，很容易与其他的冲突，所以尽可能避免正则与其他匹配方式一起使用
            //   targetString: "[0-9]*",
            //   style: const TextStyle(
            //     fontWeight: FontWeight.bold,
            //     fontSize: 30,
            //   ),
            // ),
            LVRichTextConfig(
                targetString: "+86 18868876045",
                // 如果targetString中包含特殊字符，则需要设置hasSpecialCharacters: true
                hasSpecialCharacters: true,
                style: const TextStyle(
                  decoration: TextDecoration.underline,
                )),
          ],
        ),
      ),
    );
  }
}
